File: src/main/java/model/openloopfueling/correction/OpenLoopMlhfmCorrectionManager.java
package model.openloopfueling.correction;

import contract.AfrLogFileContract;
import contract.Me7LogFileContract;
import math.Index;
import math.map.Map3d;
import model.openloopfueling.util.AfrLogUtil;
import model.openloopfueling.util.Me7LogUtil;
import org.apache.commons.math3.stat.StatUtils;
import org.apache.commons.math3.stat.descriptive.moment.Mean;

import java.util.*;

public class OpenLoopMlhfmCorrectionManager {
    private static final int LAMBDA_CONTROL_ENABLED = 0;

    private final int minPointsMe7;
    private final int minPointsAfr;
    private final double minThrottleAngle;
    private final double minRpm;
    private final double maxAfr;

    private final Map3d correctedMlhfm = new Map3d();
    private final Map<Double, List<Double>> correctionsAfrMap = new HashMap<>();

    public final Map<Double, Double> meanAfrMap = new HashMap<>();
    public final Map<Double, double[]> modeAfrMap = new HashMap<>();
    public final Map<Double, Double> correctedAfrMap = new HashMap<>();

    private OpenLoopMlhfmCorrection openLoopMlhfmCorrection;

    public OpenLoopMlhfmCorrectionManager(double minThrottleAngle, double minRpm, int minPointsMe7, int minPointsAfr, double maxAfr) {
        this.minThrottleAngle = minThrottleAngle;
        this.minRpm = minRpm;
        this.minPointsMe7 = minPointsMe7;
        this.minPointsAfr = minPointsAfr;
        this.maxAfr = maxAfr;
    }

    public void correct(Map<Me7LogFileContract.Header, List<Double>> me7Log, Map<String, List<Double>> afrLog, Map3d mlhfm) {

        List<Map<Me7LogFileContract.Header, List<Double>>> me7LogList = Me7LogUtil.findMe7Logs(me7Log, minThrottleAngle, LAMBDA_CONTROL_ENABLED, minRpm, minPointsMe7);
        List<Map<String, List<Double>>> afrLogList = AfrLogUtil.findAfrLogs(afrLog, minThrottleAngle, minRpm, maxAfr, minPointsAfr);

        generateMlhfm(mlhfm, me7LogList, afrLogList);

        openLoopMlhfmCorrection = new OpenLoopMlhfmCorrection(mlhfm, correctedMlhfm, correctedMlhfm ,correctionsAfrMap, meanAfrMap, modeAfrMap, correctedAfrMap);
    }

    public OpenLoopMlhfmCorrection getOpenLoopCorrection() {
        return openLoopMlhfmCorrection;
    }

    private void generateMlhfm(Map3d mlhfm, List<Map<Me7LogFileContract.Header, List<Double>>> me7LogList, List<Map<String, List<Double>>> afrLogList) {
        if (me7LogList.size() != afrLogList.size()) {
            throw new IllegalArgumentException("ME7 Log size does not match AFR Log size! " + me7LogList.size() + " AFR: " + afrLogList.size());
        } else if (me7LogList.size() == 0) {
            throw new IllegalArgumentException("Not enough log files! " + "ME7: " + me7LogList.size() + " AFR: " + afrLogList.size());
        }

        List<Double> mlhfmVoltage = Arrays.asList(mlhfm.yAxis);

        // Calculate the initial corrections sets
        calculateCorrections(me7LogList, afrLogList, mlhfmVoltage);
        // Process the corrections sets into a single correction
        ArrayList<Double> correctedAfrList = processCorrections(mlhfmVoltage);
        // Clean up the corrections
        postProcessCorrections(correctedAfrList);
        // Smooth the corrections
        smooth(correctedAfrList, 5);
        applyCorrections(mlhfm, correctedAfrList);
    }

    private void applyCorrections(Map3d mlhfm, ArrayList<Double> correctedAfrList) {
        Map<Double, Double> totalCorrectionError = new HashMap<>();
        List<Double> voltage = Arrays.asList(mlhfm.yAxis);

        for (int i = 0; i < voltage.size(); i++) {
            totalCorrectionError.put(voltage.get(i), correctedAfrList.get(i));
            correctedAfrMap.put(voltage.get(i), correctedAfrList.get(i));
        }

        correctedMlhfm.yAxis = voltage.toArray(new Double[0]);

        List<Double> oldKghr = new ArrayList<>();

        for(int i = 0; i < mlhfm.zAxis.length; i++) {
            oldKghr.add(mlhfm.zAxis[i][0]);
        }

        Double[][] newKghr = new Double[oldKghr.size()][1];

        for (int i = 0; i < voltage.size(); i++) {
            double oldKgPerHourValue = oldKghr.get(i);
            double totalCorrectionErrorValue = totalCorrectionError.get(voltage.get(i));

            newKghr[i][0] = (oldKgPerHourValue * ((totalCorrectionErrorValue) + 1));
        }

        correctedMlhfm.zAxis = newKghr;
    }

    private void smooth(ArrayList<Double> correctedAfrList, int window) {
        int halfWindow = window / 2;

        // Smooth
        double[] meanTempCorrections = toDoubleArray(correctedAfrList.toArray(new Double[0]));
        Mean mean = new Mean();
        for (int i = 0; i < meanTempCorrections.length; i++) {
            if (i > halfWindow && i < meanTempCorrections.length - halfWindow) {
                correctedAfrList.set(i, mean.evaluate(meanTempCorrections, i - halfWindow, window));
            }
        }
    }

    private void postProcessCorrections(ArrayList<Double> correctedAfrList) {

        // From the last known correction, fills out the list with the last known correction or 0 if a correction has not yet been found
        boolean foundStart = false;
        int lastValidCorrectionIndex = -1;
        for (int i = 0; i < correctedAfrList.size(); i++) {
            Double value = correctedAfrList.get(i);
            if (value.isNaN() && !foundStart) {
                correctedAfrList.set(i, 0d);
            } else if (!value.isNaN() && !foundStart) {
                foundStart = true;
                lastValidCorrectionIndex = i;
            } else if (value.isNaN()) {
                correctedAfrList.set(i, correctedAfrList.get(lastValidCorrectionIndex));
            } else {
                lastValidCorrectionIndex = i;
            }
        }
    }

    private ArrayList<Double> processCorrections(List<Double> mlhfmVoltage) {
        ArrayList<Double> correctedAfrList = new ArrayList<>();

        Mean mean = new Mean();
        for (Double voltage : mlhfmVoltage) {
            List<Double> corrections = correctionsAfrMap.get(voltage);
            // Get the mean of the correction set
            double meanValue = mean.evaluate(toDoubleArray(corrections.toArray(new Double[0])), 0, corrections.size());
            meanAfrMap.put(voltage, meanValue);
            // Get the mode of the correction set
            double[] mode = StatUtils.mode(toDoubleArray(corrections.toArray(new Double[0])));
            modeAfrMap.put(voltage, mode);

            // Get the average of the mean and the mode
            double correction = meanValue;
            for (double v : mode) {
                correction += v;
            }
            correction /= 1 + mode.length;

            correctedAfrList.add(correction);
        }

        return correctedAfrList;
    }

    private void calculateCorrections(List<Map<Me7LogFileContract.Header, List<Double>>> me7LogList, List<Map<String, List<Double>>> afrLogList, List<Double> mlhfmVoltageList) {
        // Loop over each log
        for (int i = 0; i < me7LogList.size(); i++) {
            Map<Me7LogFileContract.Header, List<Double>> me7Log = me7LogList.get(i);
            Map<String, List<Double>> afrLog = afrLogList.get(i);

            // For each log, loop over the voltages in MLHFM and attempt to calculate a correction
            for (int j = 0; j < mlhfmVoltageList.size(); j++) {

                double mlhfmVoltage = mlhfmVoltageList.get(j);

                correctionsAfrMap.put(mlhfmVoltage, new ArrayList<>());

                // Get the measured MAF voltages in the log
                List<Double> me7VoltageList = me7Log.get(Me7LogFileContract.Header.MAF_VOLTAGE_HEADER);

                // Attempt to find the mlhfm voltages in the log.
                List<Integer> me7VoltageIndices = getVoltageToMatchIndices(j, mlhfmVoltageList, me7VoltageList);

                // Calculate a corrected AFR for each index that is found
                for (int me7Index : me7VoltageIndices) {
                    if (me7Index != 1 && me7Index != me7VoltageList.size() - 1) {
                        double stft = me7Log.get(Me7LogFileContract.Header.STFT_COLUMN_HEADER).get(me7Index) - 1;
                        double ltft = me7Log.get(Me7LogFileContract.Header.LTFT_COLUMN_HEADER).get(me7Index) - 1;
                        double rpm = me7Log.get(Me7LogFileContract.Header.RPM_COLUMN_HEADER).get(me7Index);
                        double targetAfr = me7Log.get(Me7LogFileContract.Header.REQUESTED_LAMBDA_HEADER).get(me7Index);

                        // Find the RPM from the ME7 log in the AFR log
                        int afrIndex = Index.getInsertIndex(afrLog.get(AfrLogFileContract.RPM_HEADER), rpm);
                        List<Double> afrList = afrLog.get(AfrLogFileContract.AFR_HEADER);

                        // Now find the AFR that corresponds to the RPM in the AFR Log
                        double afr = afrList.get(Math.min(afrIndex, afrList.size() - 1)) / 14.7;

                        // Calculate a correction accounting for STFT and LTFT
                        double rawAfr = afr / (1 - (stft + ltft));
                        double afrCorrection = (rawAfr / targetAfr) - 1;

                        correctionsAfrMap.get(mlhfmVoltage).add(afrCorrection);
                    } else {
                        correctionsAfrMap.get(mlhfmVoltage).add(Double.NaN);
                    }
                }
            }
        }
    }

    private List<Integer> getVoltageToMatchIndices(int mlhfmVoltageToMatchIndex, List<Double> mlhfmVoltageList, List<Double> me7VoltageList) {
        int previousIndex = mlhfmVoltageToMatchIndex - 1;
        int nextIndex = mlhfmVoltageToMatchIndex + 1;

        double lowValue = previousIndex >= 0 ? mlhfmVoltageList.get(previousIndex) : mlhfmVoltageList.get(mlhfmVoltageToMatchIndex) - 0.0001;
        double highValue = nextIndex <= mlhfmVoltageList.size() - 1 ? mlhfmVoltageList.get(nextIndex) : mlhfmVoltageList.get(mlhfmVoltageToMatchIndex) + 0.0001;

        ArrayList<Integer> indices = new ArrayList<>();

        for (int i = 0; i < me7VoltageList.size(); i++) {
            double voltage = me7VoltageList.get(i);

            if (voltage > lowValue && voltage < highValue) {
                indices.add(i);
            }
        }

        return indices;
    }

    private double[] toDoubleArray(Double[] array) {
        double[] result = new double[array.length];

        for (int i = 0; i < result.length; i++) {
            result[i] = array[i];
        }

        return result;
    }
}

File: src/main/kotlin/domain/model/openloopfueling/correction/OpenLoopMlhfmCorrectionManager.kt
package domain.model.openloopfueling.correction

import data.contract.AfrLogFileContract
import data.contract.Me7LogFileContract
import domain.math.Index
import domain.math.map.Map3d
import domain.model.openloopfueling.util.AfrLogUtil
import domain.model.openloopfueling.util.Me7LogUtil
import org.apache.commons.math3.stat.StatUtils
import org.apache.commons.math3.stat.descriptive.moment.Mean

class OpenLoopMlhfmCorrectionManager(
    private val minThrottleAngle: Double,
    private val minRpm: Double,
    private val minPointsMe7: Int,
    private val minPointsAfr: Int,
    private val maxAfr: Double
) {
    private val correctedMlhfm = Map3d()
    private val correctionsAfrMap = mutableMapOf<Double, MutableList<Double>>()
    val meanAfrMap = mutableMapOf<Double, Double>()
    val modeAfrMap = mutableMapOf<Double, DoubleArray>()
    val correctedAfrMap = mutableMapOf<Double, Double>()

    var openLoopCorrection: OpenLoopMlhfmCorrection? = null
        private set

    fun correct(me7Log: Map<Me7LogFileContract.Header, List<Double>>, afrLog: Map<String, List<Double>>, mlhfm: Map3d) {
        val me7LogList = Me7LogUtil.findMe7Logs(me7Log, minThrottleAngle, LAMBDA_CONTROL_ENABLED, minRpm, minPointsMe7)
        val afrLogList = AfrLogUtil.findAfrLogs(afrLog, minThrottleAngle, minRpm, maxAfr, minPointsAfr)
        generateMlhfm(mlhfm, me7LogList, afrLogList)
        openLoopCorrection = OpenLoopMlhfmCorrection(mlhfm, correctedMlhfm, correctedMlhfm, correctionsAfrMap, meanAfrMap, modeAfrMap, correctedAfrMap)
    }

    private fun generateMlhfm(mlhfm: Map3d, me7LogList: List<Map<Me7LogFileContract.Header, List<Double>>>, afrLogList: List<Map<String, List<Double>>>) {
        val mlhfmVoltage = mlhfm.yAxis.toList()
        calculateCorrections(me7LogList, afrLogList, mlhfmVoltage)
        val correctedAfrList = processCorrections(mlhfmVoltage)
        postProcessCorrections(correctedAfrList)
        smooth(correctedAfrList, 5)
        applyCorrections(mlhfm, correctedAfrList)
    }

    private fun applyCorrections(mlhfm: Map3d, correctedAfrList: ArrayList<Double>) {
        val totalCorrectionError = mutableMapOf<Double, Double>()
        val voltage = mlhfm.yAxis.toList()
        for (i in voltage.indices) {
            totalCorrectionError[voltage[i]] = correctedAfrList[i]
            correctedAfrMap[voltage[i]] = correctedAfrList[i]
        }
        correctedMlhfm.yAxis = voltage.toTypedArray()
        val oldKghr = mlhfm.zAxis.map { it[0] }
        val newKghr = Array(oldKghr.size) { i ->
            arrayOf(maxOf(0.0, oldKghr[i] * (totalCorrectionError[voltage[i]]!! + 1)))
        }
        correctedMlhfm.zAxis = newKghr
    }

    private fun smooth(correctedAfrList: ArrayList<Double>, window: Int) {
        val halfWindow = window / 2
        val meanTempCorrections = correctedAfrList.map { it }.toDoubleArray()
        val mean = Mean()
        for (i in meanTempCorrections.indices) {
            if (i > halfWindow && i < meanTempCorrections.size - halfWindow) {
                correctedAfrList[i] = mean.evaluate(meanTempCorrections, i - halfWindow, window)
            }
        }
    }

    private fun postProcessCorrections(correctedAfrList: ArrayList<Double>) {
        var foundStart = false
        var lastValidCorrectionIndex = -1
        for (i in correctedAfrList.indices) {
            val value = correctedAfrList[i]
            when {
                value.isNaN() && !foundStart -> correctedAfrList[i] = 0.0
                !value.isNaN() && !foundStart -> { foundStart = true; lastValidCorrectionIndex = i }
                value.isNaN() -> correctedAfrList[i] = correctedAfrList[lastValidCorrectionIndex]
                else -> lastValidCorrectionIndex = i
            }
        }
    }

    private fun processCorrections(mlhfmVoltage: List<Double>): ArrayList<Double> {
        val correctedAfrList = arrayListOf<Double>()
        val mean = Mean()
        for (voltage in mlhfmVoltage) {
            val corrections = correctionsAfrMap[voltage]!!
            val meanValue = mean.evaluate(corrections.toDoubleArray(), 0, corrections.size)
            meanAfrMap[voltage] = meanValue
            val mode = StatUtils.mode(corrections.toDoubleArray())
            modeAfrMap[voltage] = mode
            var correction = meanValue
            for (v in mode) correction += v
            correction /= 1 + mode.size
            correctedAfrList.add(correction)
        }
        return correctedAfrList
    }

    private fun calculateCorrections(me7LogList: List<Map<Me7LogFileContract.Header, List<Double>>>, afrLogList: List<Map<String, List<Double>>>, mlhfmVoltageList: List<Double>) {
        val size = minOf(me7LogList.size, afrLogList.size)
        if (me7LogList.size != afrLogList.size) {
            System.err.println("[OpenLoopMlhfmCorrectionManager] Warning: ME7 pull count (${me7LogList.size}) does not match AFR pull count (${afrLogList.size}). Only the first $size pulls will be used.")
        }

        for (mlhfmVoltage in mlhfmVoltageList) {
            correctionsAfrMap[mlhfmVoltage] = mutableListOf()
        }

        for (i in 0 until size) {
            val me7Log = me7LogList[i]
            val afrLog = afrLogList[i]

            for (j in mlhfmVoltageList.indices) {
                val mlhfmVoltage = mlhfmVoltageList[j]
                val me7VoltageList = me7Log[Me7LogFileContract.Header.MAF_VOLTAGE_HEADER]!!
                val me7VoltageIndices = getVoltageToMatchIndices(j, mlhfmVoltageList, me7VoltageList)

                for (me7Index in me7VoltageIndices) {
                    if (me7Index != 0 && me7Index != me7VoltageList.size - 1) {
                        val stft = me7Log[Me7LogFileContract.Header.STFT_COLUMN_HEADER]!![me7Index] - 1
                        val ltft = me7Log[Me7LogFileContract.Header.LTFT_COLUMN_HEADER]!![me7Index] - 1
                        val rpm = me7Log[Me7LogFileContract.Header.RPM_COLUMN_HEADER]!![me7Index]
                        val targetAfr = me7Log[Me7LogFileContract.Header.REQUESTED_LAMBDA_HEADER]!![me7Index]

                        val afrIndex = Index.getInsertIndex(afrLog[AfrLogFileContract.RPM_HEADER]!!, rpm)
                        val afrList = afrLog[AfrLogFileContract.AFR_HEADER]!!
                        val afr = afrList[minOf(afrIndex, afrList.size - 1)] / 14.7
                        val rawAfr = afr / (1 - (stft + ltft))
                        val afrCorrection = (rawAfr / targetAfr) - 1

                        correctionsAfrMap[mlhfmVoltage]!!.add(afrCorrection)
                    } else {
                        correctionsAfrMap[mlhfmVoltage]!!.add(Double.NaN)
                    }
                }
            }
        }
    }

    private fun getVoltageToMatchIndices(mlhfmVoltageToMatchIndex: Int, mlhfmVoltageList: List<Double>, me7VoltageList: List<Double>): List<Int> {
        val previousIndex = mlhfmVoltageToMatchIndex - 1
        val nextIndex = mlhfmVoltageToMatchIndex + 1

        val lowValue = if (previousIndex >= 0) mlhfmVoltageList[previousIndex] else mlhfmVoltageList[mlhfmVoltageToMatchIndex] - 0.0001
        val highValue = if (nextIndex <= mlhfmVoltageList.size - 1) mlhfmVoltageList[nextIndex] else mlhfmVoltageList[mlhfmVoltageToMatchIndex] + 0.0001

        return me7VoltageList.indices.filter { me7VoltageList[it] in (lowValue + Double.MIN_VALUE)..highValue }
    }

    companion object {
        private const val LAMBDA_CONTROL_ENABLED = 0.0
    }
}
